### PR TITLE
allow getting recoverable_connection_errors without an active transport 

### PR DESCRIPTION
**Pull request by pawl**
I was looking into solutions for the [celery + redis memory leak issue](https://github.com/celery/celery/issues/4843#issuecomment-999168967) and I was attempting to reduce the references to initialized `Transport`s.

One thing I noticed was that you currently need at least one initialized `Transport` with an active connection to allow filling the cache for `recoverable_connection_errors`/`recoverable_channel_errors`/`connection_errors`/`channel_errors`. It seems like it makes more sense to get the errors from the `Transport` class without needing to initialize it (in case there are connection errors on the first `Transport` initialization).

This doesn't fix the memory leak issue, but maybe this change is still worth making?

### CODE REVIEW

File: kombu/transport/consul.py
Comment (auvipy) [inline kombu/transport/consul.py:279]: only 3 transports are subject to this adjustments?

File: t/unit/test_connection.py
Comment (auvipy) [inline t/unit/test_connection.py:576]: beside the unit tests I would love to see integration tests for this

**Review by auvipy — CHANGES_REQUESTED** [review]

File: kombu/transport/consul.py
Comment (pawl) [reply kombu/transport/consul.py:279]: Yeah, I did a search for `self.connection_errors` and `self.channel_errors`, and looked at each of the transports.

For some reason, the librabbitmq and mongodb backends don't have the `try/except ImportError:` logic and they don't use `__init__` to fill in `connection_errors` and `channel_errors`.

**Review by auvipy — CHANGES_REQUESTED** [review]
with pytest.raises(KeyError):
            self.channel._receive_one(self.channel.subclient)
>       assert not self.channel._in_listen
E       assert not True
E        +  where True = <t.unit.transport.test_redis.Channel object at 0x7f8bf8d594d0>._in_listen
E        +    where <t.unit.transport.test_redis.Channel object at 0x7f8bf8d594d0> = <t.unit.transport.test_redis.test_Channel object at 0x7f8bf8d591d0>.channel

t/unit/transport/test_redis.py:582: AssertionError

**Review by auvipy — CHANGES_REQUESTED** [review]
I would like to take some time & more eyes and trials before merging this. in the mean the you can push any improvements in mind. and most possibly the more integration tests the better IMHO

**Comment by auvipy** [comment]
besides, do we need any docs changes to mention the adjustments being made?

**Comment by auvipy** [comment]
@ michael-lazar

**Comment by auvipy** [comment]
@pawl did you try this in any project with celery? i want to have some integration test for this

**Comment by pawl** [comment]
@auvipy I added an integration test here: https://github.com/celery/kombu/pull/1471/commits/448cdb9a4bf3364eef4025da69f339e954b116ea

> do we need any docs changes to mention the adjustments being made?

Oops, should I be updating the changelog?

I'm not aware of any documentation changes since old code should continue to work the same way, because the errors are still a class attribute on the same class.

I'm not sure we need to document the removal of `_get_errors` outside of the release notes since it's an internal method. It also doesn't look like it's overridden in any open source projects: https://github.com/search?q=transport+_get_errors+NOT+MultiChannelPoller+NOT+test_purge+NOT+%22by+Ask+Solem%22&type=Code

**Comment by pawl** [comment]
> did you try this in any project with celery?

I've tested the change with [this branch](https://github.com/pawl/celery_pyamqp_memory_leak/tree/test_errors_fix) of my celery memory leak example repo. It's still able to recover from connection errors successfully after I restart redis. Also, it connects successfully on start-up.

**Comment by auvipy** [comment]
ok we are good to go

**Review by auvipy — APPROVED** [review]

**Comment by auvipy** [comment]
decided to land this change on 5.3 for now, as it changes some behaviours, but if other core team members agree will end landing in 5.2.x bug fix release

File: kombu/connection.py
Comment (matusvalo) [inline kombu/connection.py:951]: Is it correct to put parameter of `get_transport_cls()` as None?

File: kombu/connection.py
Comment (matusvalo) [inline kombu/connection.py:958]: Here the same.

File: kombu/connection.py
Comment (matusvalo) [inline kombu/connection.py:963]: Here the same.

**Review by matusvalo — CHANGES_REQUESTED** [review]
I am not 100% sure but if I read correctly the code `get_transport_cls()` needs a transport as string or type. Default value is `None` which I am not sure will work.

https://github.com/celery/kombu/blob/b6b44085755c121fbcc8d954bd8f62ebe56ebc12/kombu/transport/__init__.py#L47-L86

**Comment by pawl** [comment]
@matusvalo `get_transport_cls` is already used without arguments in `create_transport` (used by the `transport` property): https://github.com/celery/kombu/blob/a9c4f9837c1ad2296a9b21311f79c8cdbddade6b/kombu/connection.py#L606

It looks like `_init_params` will populate `self.transport_cls` and it will be used by `get_transport_cls`.

**Comment by matusvalo** [comment]
> @matusvalo `get_transport_cls` is already used without arguments in `create_transport` (used by the `transport` property):
> 
> https://github.com/celery/kombu/blob/a9c4f9837c1ad2296a9b21311f79c8cdbddade6b/kombu/connection.py#L606
> 
> It looks like `_init_params` will populate `self.transport_cls` and it will be used by `get_transport_cls`.

Yeah you are right :-/. I missed that it is wrapped in the same class. Please ignore my comment.

**Review by matusvalo — APPROVED** [review]

**Pull request merged**
